File: js/SequenceSuite.js
const Sequence = require('./Sequence.js').Sequence;
const NaturalnessModel = require('./NaturalnessModel.js').NaturalnessModel;

const DEPTH = 4;
const PROBA_OF_UNKNOWN = 0;//0.000001;
const UPPER_BOUND = -Math.log2(PROBA_OF_UNKNOWN);

class SequenceSuite {
    constructor(suite, depth, probaOfUnknown) {
        this.depth = depth || DEPTH;
        this.probaOfUnknown = probaOfUnknown || PROBA_OF_UNKNOWN;
        if (suite === null || suite === undefined) {
            throw 'cannot create SequenceSuite with no suite';
        }
        if (!(Array.isArray(suite))) {
            throw 'cannot create SequenceSuite with no array';
        }
        suite.forEach(seq => {
            if (!(seq instanceof Sequence)) {
                throw 'cannot create SequenceSuite with no array of Sequence';
            }
        })
        this.suite = suite;
    }

    getMoreNatural() {
        if (this.suite.length === 0) {
            throw 'Suite is empty';
        }

        if (this.suite.length === 1) {
            return {
                crossEntropy : UPPER_BOUND,
                sequence : this.suite[0]
            }
        }
        let moreNatural = {
            crossEntropy : UPPER_BOUND
        }
        this.suite.forEach( (sequence, index) => {
            let model = new NaturalnessModel(this.depth, this.probaOfUnknown);
            let suiteCopy = this.suite.slice(0);
            suiteCopy.splice(index, 1);
            suiteCopy.forEach(seq => model.learn(seq));
            let crossEntropy = model.crossEntropy(sequence);
            if (crossEntropy <= moreNatural.crossEntropy) {
                moreNatural.crossEntropy = crossEntropy;
                moreNatural.sequence = sequence;
                moreNatural.index = index;
            }
        });
        return moreNatural;
    }

    rank() {
        if (this.suite.length === 0) return [];

        let rank = [];
        let moreNatural = this.getMoreNatural();
        if (moreNatural.sequence === undefined) console.log('undefined!!!');
        rank.push({sequence:moreNatural.sequence, crossEntropy:moreNatural.crossEntropy});
        let suiteCopy = this.suite.slice(0);
        suiteCopy.splice(moreNatural.index, 1);
        let other = new SequenceSuite(suiteCopy, this.depth, this.probaOfUnknown);
        rank.push(...other.rank());
        return rank;
    }
}

module.exports.SequenceSuite = SequenceSuite;

File: java/src/main/java/com/naturalness/SequenceSuite.java
/*
Licensed to the Apache Software Foundation (ASF) under one
or more contributor license agreements.  See the NOTICE file
distributed with this work for additional information
regarding copyright ownership.  The ASF licenses this file
to you under the Apache License, Version 2.0 (the
"License"); you may not use this file except in compliance
with the License.  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing,
software distributed under the License is distributed on an
"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, either express or implied.  See the License for the
specific language governing permissions and limitations
under the License.
*/

package com.naturalness;

import java.util.List;
import java.util.ArrayList;

public class SequenceSuite {
    private final int DEPTH_DEFAULT = 3;
    private final double PROBA_OF_UNKNOWN_DEFAULT = 1e-6;

    private int depth;
    private double probaOfUnknown;
    private List<Sequence> sequenceList;

    public SequenceSuite(List<Sequence> sequenceList) {
        if (sequenceList == null) {
            throw new IllegalArgumentException("SequenceSuite new with null");
        }
        this.depth = DEPTH_DEFAULT;
        this.probaOfUnknown = PROBA_OF_UNKNOWN_DEFAULT;
        this.sequenceList = new ArrayList<Sequence>(sequenceList);
    }

    public SequenceSuite(List<Sequence> sequenceList, int depth, double probaOfUnknown) {
        this(sequenceList);
        this.depth = depth;
        this.probaOfUnknown = probaOfUnknown;
    }

    public Ranking getMoreNatural() {
        if (sequenceList.size() == 0) {
            throw new IllegalStateException("SequenceSuite is empty");
        }

        double minEntropy = -Math.log(probaOfUnknown)/Math.log(2);
        Sequence minSequence = sequenceList.get(0);

        if (sequenceList.size() == 1) {
            return new Ranking(sequenceList.get(0), minEntropy);
        }

        for (int i = 0; i < sequenceList.size(); i++) {
            NaturalnessModel model = new NaturalnessModel(this.depth, this.probaOfUnknown);
            List<Sequence> suiteCopy = new ArrayList<Sequence>(sequenceList);
            Sequence current = suiteCopy.remove(i);
            for (Sequence sequence : suiteCopy) {
                model.learn(sequence);
            }
            double currentCrossEntropy = model.crossEntropy(current);
            if (currentCrossEntropy < minEntropy) {
                minEntropy = currentCrossEntropy;
                minSequence = current;
            }
        }
        return new Ranking(minSequence, minEntropy);
    }

    public List<Ranking> rank() {
        if (sequenceList.size() == 0) {
            return new ArrayList<Ranking>();
        }
        List<Ranking> rankingList = new ArrayList<Ranking>();
        Ranking moreNatural = getMoreNatural();
        rankingList.add(moreNatural);
        List<Sequence> suiteCopy = new ArrayList<Sequence>(sequenceList);
        suiteCopy.remove(moreNatural.getSequence());
        SequenceSuite other = new SequenceSuite(suiteCopy);
        rankingList.addAll(other.rank());
        return rankingList;
    }
}